managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/cache/RangeEntryCacheImpl.java

```diff
@@ -295,51 +439,70 @@ private void asyncReadEntry0(ReadHandle lh, long firstEntry, long lastEntry, boo
+                                        requireNonNull(ml.getName());
+                                        requireNonNull(ml.getExecutor());
+
+                                        try {
+                                            // We got the entries, we need to transform them to a List<> type
+                                            long totalSize = 0;
+                                            final List<EntryImpl> entriesToReturn =
+                                                    Lists.newArrayListWithExpectedSize(entriesToRead);
+                                            for (LedgerEntry e : ledgerEntries) {
+                                                EntryImpl entry = RangeEntryCacheManagerImpl.create(e, interceptor);
+                                                entriesToReturn.add(entry);
+                                                totalSize += entry.getLength();
+                                                if (shouldCacheEntry) {
+                                                    EntryImpl cacheEntry = EntryImpl.create(entry);
+                                                    insert(cacheEntry);
+                                                    cacheEntry.release();
+                                                }
+                                            }
+
+                                            manager.mlFactoryMBean.recordCacheMiss(entriesToReturn.size(), totalSize);
+                                            ml.getMbean().addReadEntriesSample(entriesToReturn.size(), totalSize);
+
+                                            return entriesToReturn;
+                                        } finally {
+                                            ledgerEntries.close();
+                                        }
+                                    });
+                    // handle LH invalidation
+                    readResult.exceptionally(exception -> {
                         if (exception instanceof BKException
                                 && ((BKException) exception).getCode() == BKException.Code.TooManyRequestsException) {
-                            callback.readEntriesFailed(createManagedLedgerException(exception), ctx);
                         } else {
                             ml.invalidateLedgerHandle(lh);
-                            ManagedLedgerException mlException = createManagedLedgerException(exception);
-                            callback.readEntriesFailed(mlException, ctx);
                         }
                         return null;
-            });
+                    });
+
+                    cachedPendingRead.attach(readResult);
+                }
+            }
         }
     }
 
     @Override
     public void clear() {
         Pair<Integer, Long> removedPair = entries.clear();
         manager.entriesRemoved(removedPair.getRight(), removedPair.getLeft());
+        cachedPendingReads.clear();
```

Comment:
I noticed the element of the `cachedPendingReads` only be removed here. This one will be called if all the cached data should be clean up

It's better to add a unit test for this one to make sure we will not introduce any heap memory leak.

Reply:
the elements here are one entry per ledger.
the per ledger maps are evicted here
https://github.com/apache/pulsar/pull/17241/files#diff-c55509e3ab1389d89a58fd564f2e318dbb95f50121ab33c729a7ca4a21d02ef1R340

We could remove the entry for a ledger in case of rollover.
but I thought it requires more coordination.
I will try to improve this

Reply:
> We could remove the entry for a ledger in case of rollover.
but I thought it requires more coordination.
I will try to improve this

Yes, we are on the same page.